### ISSUE TITLE
[FR] TrainingClient 缺少 checkpoint 管理方法：save_state / load_state / load_state_with_optimizer

### ISSUE THREAD

**Issue by void-main** [nex-agi/weaver#28]
## 背景

weaver-server 已经实现了完整的 checkpoint 管理接口：

- `POST /models/:id/checkpoints` — 保存 checkpoint
- `POST /models/:id/load` — 从 checkpoint 恢复
- `GET /models/:id/checkpoints` — 列出 checkpoints
- `GET /checkpoints/:id` — 获取单个 checkpoint
- `DELETE /checkpoints/:id` — 删除 checkpoint

但 **`TrainingClient` 完全没有暴露这些能力**，用户无法通过 SDK 做断点续训。

对比 Tinker（Thinking Machines Lab），checkpoint 管理是一等公民：

```python
# 保存 checkpoint
path = training_client.save_state(name="iter-100")

# 恢复（仅权重，Adam momentum 重置）
training_client.load_state(path)

# 恢复（权重 + Adam momentum，真正的断点续训）
training_client.load_state_with_optimizer(path)
```

---

### `TrainingClient.save_state()`

```python
def save_state(
    self,
    *,
    name: str | None = None,
    checkpoint_type: str = "training",
    wait: bool = True,
) -> str:
    """保存当前模型权重为 checkpoint，返回 checkpoint path。

    Args:
        name: checkpoint 名称（可选，服务端自动生成）
        checkpoint_type: "training"（默认）或 "training_with_optimizer"
        wait: True 则等待完成后返回 path，False 返回 OperationHandle

    Returns:
        Checkpoint path（如 "weaver://run-id/weights/checkpoint-001"）
    """
```

内部调用：`POST /models/:id/checkpoints`

---

### `TrainingClient.load_state()`

```python
def load_state(
    self,
    path: str,
    *,
    wait: bool = True,
) -> OperationHandle | dict:
    """从 checkpoint 恢复模型权重（不恢复 optimizer state）。

    Args:
        path: checkpoint path（来自 save_state() 的返回值，或 list_checkpoints()）
        wait: True 则等待完成
    """
```

内部调用：`POST /models/:id/load`，`include_optimizer=False`

---

### `TrainingClient.load_state_with_optimizer()`

```python
def load_state_with_optimizer(
    self,
    path: str,
    *,
    wait: bool = True,
) -> OperationHandle | dict:
    """从 checkpoint 恢复模型权重 + Adam optimizer state（断点续训）。

    Args:
        path: checkpoint path
        wait: True 则等待完成
    """
```

内部调用：`POST /models/:id/load`，`include_optimizer=True`

---

### `TrainingClient.list_checkpoints()`（可选，锦上添花）

```python
def list_checkpoints(self) -> list[dict]:
    """列出该模型的所有 checkpoints。"""
```

内部调用：`GET /models/:id/checkpoints`

---

## 注意

- `load_state_with_optimizer` 依赖 weaver-server 的 `POST /models/:id/load` 增加 `include_optimizer` 参数，以及 trainer 侧支持 save/load Adam momentum state（需联合 weaver-server 一起确认，见 [weaver-server#74](https://github.com/china-qijizhifeng/weaver-server/issues/74)）
- `save_state()` 中 `checkpoint_type="training_with_optimizer"` 同理，依赖 server 支持
- `load_state()` 仅需要 server 现有接口，可以先实现

## 相关

- weaver-server checkpoint 接口：`internal/http/handlers/weights.go`
- Tinker 参考：https://tinker-docs.thinkingmachines.ai/api-reference/trainingclient
- weaver-server gap tracking：china-qijizhifeng/weaver-server#74

**Comment by void-main** [comment]
## 测试方法

基于 `examples/pig_latin.py` 的代码风格，写了一个最小化的验证脚本，覆盖三个新方法的核心行为。

**运行方式**
```bash
WEAVER_API_KEY=sk-... python checkpoint_test.py
```

**测试逻辑**

1. 热身训练 3 步，记录此时的 loss（`loss_before`）
2. 调 `save_state()` 保存 checkpoint，断言返回非空 path
3. 再训练 3 步，让权重偏离 checkpoint（`loss_diverged`）
4. 调 `load_state(ckpt_path)` 恢复
5. 重算 loss，断言其与 `loss_before` 差值 < 0.05（即权重确实回到了 checkpoint 状态）
6. 调 `save_state(checkpoint_type="training_with_optimizer")` 保存带 optimizer state 的 checkpoint
7. 再训练偏离，调 `load_state_with_optimizer()` 恢复，验证无报错（optimizer state 正确性通过后续收敛速度验证，脚本只做基础 smoke test）

**完整脚本**

```python
"""
Minimal test script for TrainingClient checkpoint methods:
  - save_state()
  - load_state()
  - load_state_with_optimizer()

Run:
    WEAVER_API_KEY=sk-... python checkpoint_test.py
"""

from __future__ import annotations

import os
import torch
from weaver import ServiceClient, types

BASE_MODEL = "Qwen/Qwen3-8B"

EXAMPLES = [
    {"input": "banana split", "output": "anana-bay plit-say"},
    {"input": "rubber duck",  "output": "ubber-ray uck-day"},
]


def make_datum(example: dict, tokenizer) -> types.Datum:
    prompt = f"English: {example['input']}\nPig Latin:"
    prompt_tokens = tokenizer.encode(prompt, add_special_tokens=True)
    completion_tokens = tokenizer.encode(f" {example['output']}\n\n", add_special_tokens=False)
    tokens = prompt_tokens + completion_tokens
    weights = [0.0] * len(prompt_tokens) + [1.0] * len(completion_tokens)
    return types.Datum(
        model_input=types.ModelInput.from_ints(tokens[:-1]),
        loss_fn_inputs={
            "target_tokens": torch.tensor(tokens[1:], dtype=torch.int64),
            "weights": torch.tensor(weights[1:], dtype=torch.float32),
        },
    )


def compute_loss(fwdbwd_result: dict, data: list) -> float:
    outputs = fwdbwd_result.get("result", {}).get("loss_fn_outputs") or []
    logprobs = torch.cat(
        [torch.as_tensor(o.get("logprobs") or o.get("Logprobs"), dtype=torch.float32)
         for o in outputs]
    )
    weights = torch.cat([d.loss_fn_inputs["weights"] for d in data])
    return float(-torch.dot(logprobs, weights) / weights.sum())


def train_steps(client, data, adam, n: int, label: str) -> float:
    loss = 0.0
    for i in range(n):
        result = client.forward_backward(data, "cross_entropy", wait=True)
        client.optim_step(adam, wait=True)
        loss = compute_loss(result, data)
        print(f"  [{label}] step {i+1}: loss={loss:.4f}")
    return loss


def main():
    with ServiceClient(api_key=os.getenv("WEAVER_API_KEY")) as svc:
        client = svc.create_model(base_model=BASE_MODEL)
        tokenizer = client.get_tokenizer()
        data = [make_datum(e, tokenizer) for e in EXAMPLES]
        adam = types.AdamParams(learning_rate=1e-4)

        # Phase 1: 热身 3 步
        print("\n[Phase 1] Warm-up training (3 steps)")
        loss_before = train_steps(client, data, adam, 3, "warmup")

        # Phase 2: save_state（新接口）
        print("\n[Phase 2] save_state()")
        ckpt_path = client.save_state(name="test-ckpt")
        print(f"  checkpoint path: {ckpt_path}")
        assert ckpt_path, "save_state() should return a non-empty path"

        # Phase 3: 继续训练 3 步偏离
        print("\n[Phase 3] Continue training (3 more steps, diverge from checkpoint)")
        loss_diverged = train_steps(client, data, adam, 3, "diverge")

        # Phase 4: load_state（新接口）
        print("\n[Phase 4] load_state() — restore weights only")
        client.load_state(ckpt_path)

        result_after_load = client.forward_backward(data, "cross_entropy", wait=True)
        loss_after_load = compute_loss(result_after_load, data)
        print(f"  loss before save:    {loss_before:.4f}")
        print(f"  loss after diverge:  {loss_diverged:.4f}")
        print(f"  loss after load:     {loss_after_load:.4f}")

        diff = abs(loss_after_load - loss_before)
        assert diff < 0.05, (
            f"After load_state(), loss {loss_after_load:.4f} should be close to "
            f"pre-save loss {loss_before:.4f} (diff={diff:.4f})"
        )
        print("  ✓ load_state() verified: loss matches checkpoint")

        # Phase 5: save_state with optimizer
        print("\n[Phase 5] save_state(checkpoint_type='training_with_optimizer')")
        ckpt_with_opt = client.save_state(
            name="test-ckpt-with-opt",
            checkpoint_type="training_with_optimizer",
        )
        print(f"  checkpoint path: {ckpt_with_opt}")

        train_steps(client, data, adam, 3, "diverge-again")

        # Phase 6: load_state_with_optimizer（新接口）
        print("\n[Phase 6] load_state_with_optimizer()")
        client.load_state_with_optimizer(ckpt_with_opt)

        result_opt = client.forward_backward(data, "cross_entropy", wait=True)
        loss_opt = compute_loss(result_opt, data)
        print(f"  loss after load_with_optimizer: {loss_opt:.4f}")
        print("  ✓ load_state_with_optimizer() completed without error")

        print("\n✅ All checkpoint tests passed.")


if __name__ == "__main__":
    main()
```

**期望输出（示意）**

```
[Phase 1] Warm-up training (3 steps)
  [warmup] step 1: loss=3.1234
  [warmup] step 2: loss=2.9876
  [warmup] step 3: loss=2.8543

[Phase 2] save_state()
  checkpoint path: weaver://xxx/weights/test-ckpt

[Phase 3] Continue training (3 more steps, diverge from checkpoint)
  [diverge] step 1: loss=2.7123
  [diverge] step 2: loss=2.5891
  [diverge] step 3: loss=2.4602

[Phase 4] load_state() — restore weights only
  loss before save:    2.8543
  loss after diverge:  2.4602
  loss after load:     2.8551   ← 回到了 checkpoint 时的水平
  ✓ load_state() verified: loss matches checkpoint

...

✅ All checkpoint tests passed.
```
